File: py/kaleido.py
"""Dihedral kaleidoscope transform — see spec/kaleido.md.
Mirrors rs/src/kaleido.rs exactly.
"""

from __future__ import annotations
import math
import numpy as np
from sample import sample_bilinear


def kaleido(
    source: np.ndarray,
    segments: int,
    rotation_deg: float,
    zoom: float,
) -> np.ndarray:
    """Apply dihedral kaleidoscope to source.

    source       : (H, W, 4) float32 linear RGBA
    segments     : number of dihedral wedges (2–24)
    rotation_deg : pattern rotation in degrees (0–360)
    zoom         : >1 samples a smaller central source region (0.25–4.0)
    Returns      : (H, W, 4) float32 linear RGBA (new array)
    """
    h, w = source.shape[:2]
    cx = (w - 1) / 2.0
    cy = (h - 1) / 2.0
    rotation = rotation_deg * math.pi / 180.0
    wedge_w = 2.0 * math.pi / segments

    # Build f64 coordinate grids
    xs = np.arange(w, dtype=np.float64)
    ys = np.arange(h, dtype=np.float64)
    xx, yy = np.meshgrid(xs, ys)      # (H, W)

    dx = xx - cx
    dy = yy - cy
    r  = np.hypot(dx, dy)
    a  = np.arctan2(dy, dx) + rotation   # float64

    # --- Dihedral fold ---
    k_f = np.floor(a / wedge_w)          # float64, integer values
    am  = a - k_f * wedge_w              # residual in [0, wedge_w]
    am  = np.clip(am, 0.0, wedge_w)      # fp guard at exact multiples

    # Odd wedge → mirror; use int64 for correct negative-k modulo
    k_int  = k_f.astype(np.int64)
    mirror = (k_int % 2) == 1
    am     = np.where(mirror, wedge_w - am, am)

    # --- Source coordinates ---
    sx = cx + (r / zoom) * np.cos(am)
    sy = cy + (r / zoom) * np.sin(am)

    # --- Bilinear sample (edge-clamped, f64 arithmetic) ---
    sampled = sample_bilinear(source, sx, sy)   # (H, W, 4) float64
    return np.clip(sampled, 0.0, 1.0).astype(np.float32)

File: rs/src/kaleido.rs
//! Dihedral kaleidoscope transform — see spec/kaleido.md.
//! Mirrors py/kaleido.py exactly.

use crate::layers::{BlendMode, Layer};
use crate::sample::sample_bilinear;
use std::f64::consts::{PI, TAU};

/// Apply dihedral kaleidoscope to `source`, return a new layer.
///
/// * `segments`     — number of dihedral wedges (2–24)
/// * `rotation_deg` — pattern rotation in degrees (0–360)
/// * `zoom`         — >1 samples a smaller central source region (0.25–4.0)
pub fn kaleido(source: &Layer, segments: u32, rotation_deg: f64, zoom: f64) -> Layer {
    let w  = source.width  as usize;
    let h  = source.height as usize;
    let cx = (w as f64 - 1.0) / 2.0;
    let cy = (h as f64 - 1.0) / 2.0;
    let rotation = rotation_deg * PI / 180.0;
    let wedge_w  = TAU / segments as f64;

    let mut out_rgba = vec![0.0f32; w * h * 4];

    for y in 0..h {
        for x in 0..w {
            let dx = x as f64 - cx;
            let dy = y as f64 - cy;
            let r  = dx.hypot(dy);
            let a  = dy.atan2(dx) + rotation;

            // --- Dihedral fold ---
            let k  = (a / wedge_w).floor();
            let am = (a - k * wedge_w).clamp(0.0, wedge_w);
            let am_fold = if (k as i64).rem_euclid(2) == 1 {
                wedge_w - am
            } else {
                am
            };

            // --- Source coordinates ---
            let sx = cx + (r / zoom) * am_fold.cos();
            let sy = cy + (r / zoom) * am_fold.sin();

            // --- Bilinear sample (edge-clamped, f64 arithmetic) ---
            let samp = sample_bilinear(&source.rgba, w, h, sx, sy);
            let i = (y * w + x) * 4;
            out_rgba[i    ] = samp[0] as f32;
            out_rgba[i + 1] = samp[1] as f32;
            out_rgba[i + 2] = samp[2] as f32;
            out_rgba[i + 3] = samp[3] as f32;
        }
    }

    Layer {
        rgba:    out_rgba,
        width:   source.width,
        height:  source.height,
        mode:    BlendMode::Normal,
        opacity: 1.0,
        visible: true,
        name:    "kaleido".into(),
    }
}
